text/0000-geoip-plugins.md

```diff
@@ -0,0 +1,140 @@
+banner. Other potential use cases include CMS sites showing different content
+based on the user's country or map applications wanting to show a relevant
+section of the world to each user.
+
+The end goal is to have plugins that site builders can install, configure with
+a Maxmind database file, and immediately start using geoip in their Fusion
+application.
+	
+## Detailed design
+
+### fusion-plugin-geoip
+
+We will use `@lxe/maxmind-db-reader` to load the geoip data from `ctx.ip` on
+the server. A `geoip data` is a specific object that we define here using a
+Flow type. The Maxmind DB reader needs to be initialized with a Maxmind .mmdb
+file, so we additionally define a `GeoipConfigToken`. The types are as follows:
+
+```js
+type GeoipConfig = {
+  dbPath: string,
+  defaultGeoData?: GeoipData
+};
+
+type GeoipData = {
+  continentCode: string,
+  continentName: string,
+  countryCode: string,
+  countryName: string,
+  cityName: ?string,
+  lat: number,
+  lng: number,
+  timeZone: string, // e.g., "America/Los_Angeles"
+};
+```
+
+The type of `GeoIP.from` is `(ctx: Context) => GeoipData` and it will be
+memoized. The `dbPath` is required and the server should not start if it is not
+a valid path to a .mmdb database file. All consumers use `GeoIP.from` to load
+geoip data from the context. If geoip data is not available, then the user has
+the option to configure `defaultGeoData` that will be used instead. If geoip
+data is not available and the user does not configure `defaultGeoData`, geoip
+will resolve to a default value chosen by the package. This is acceptable
+because GeoIP is generally understood to be a best-effort estimation and can
+never achieve 100% accuracy.
+
+We will additionally define a middleware that includes the geoip data in JSON
+in a script tag on the page. We will then load this tag in the browser in
+order to provide geoip there. This design is similar to how
+fusion-plugin-i18n works.
+
+### fusion-plugin-geoip-react
+
+We will set up a context provider that provides GeoIP data, and a `withGeoip`
+HOC that consumes the geoip data from React context. The HOC will inject the
+`geoip` prop containing geoip data.
+
+## Drawbacks
+
+The fusion team would have to support two more plugins. The feature could be
+implemented in user space, albeit with significant duplication. The impact on
+teaching people Fusion should be small, since geoip is a relatively advanced
+feature and probably will not be included in create-fusion-app. There is no
+cost to migrate existing Fusion applications to it, unless they have already
+implemented their own geoip, in which case they would just need to swap out a
+plugin. It does not have any integration points with existing open-source
+fusion packages, which helps reduce complexity.
+
+The trickiest part of owning this package will probably be owning the setup and
+API documentation. We will need to clearly document that users need to bring
+their own Maxmind database file.
```

Comment:
How much work is generating one of these? Could one be provided out of the box? This plugin could swing between being useful to unusable depending on how difficult this is to setup.

Reply:
It looks like we could bundle the plugin with a [GeoLite2](https://dev.maxmind.com/geoip/geoip2/geolite2/) database if we need to make it work out of the box. This is a smaller, less accurate free version of their GeoIP database product. The full product requires a license and subscription to use.